### ISSUE TITLE
Don't increment the for loop variable when there's a break statement

### ISSUE THREAD

**Issue by kchristin22** [vgvassilev/clad#851]
Due to PR #833, the step of the for loop variable is performed for every branch case of the for loop. However, when there's a break statement in the for loop, the step is not applied afterwards, which isn't the case when there's a continue statement or neither of them. So we shouldn't increment the variable beforehand in this case.

Example:
```cpp
#include "clad/Differentiator/Differentiator.h"
#include <iostream>

#define show(x) std::cout << #x << ": " << x << "\n";

double fn(double u, double v) {
    double res = 0;
    for (int i = 1; i < 3; i++) {
        res += i * u;
        if (i == 1)
            break;
    }
    return res;
}

int main() {
    auto d_fn = clad::gradient(fn);
    double u = 3, v = 5;
    double du, dv;
    du = dv = 0;
    d_fn.execute(u, v, &du, &dv);
    show(du);
    show(dv);
}
```
Expected output:
**du** = 3
**dv** = 0

Output:
**du** = 0
**dv** = 0

**Comment by kchristin22** [comment]
I have a branch ready to merge for this issue once the PR #833 is merged

**Comment by ovdiiuv** [comment]
Well, the provided example seems to work:
`#include "clad/Differentiator/Differentiator.h"
#include <iostream>

#define show(x) std::cout << #x << ": " << x << "\n";

double fn(double u, double v) {
    double res = 0;
    for (int i = 1; i < 3; i++) {
        res += i * u;
        if (i == 1)
            break;
    }
    return res;
}

int main() {
    auto d_fn = clad::gradient(fn);
    double u = 3, v = 5;
    double du, dv;
    du = dv = 0;
    d_fn.execute(u, v, &du, &dv);
    show(du);
    show(dv);
}`

the expected output is **du=1**, **dv=0** tho and it matches the output. I think the issue isn't relevant and we can close it

**Comment by vgvassilev** [comment]
@parth-07 what do you think?

**Comment by kchristin22** [comment]
Hi! Yes, this currently works. It's mentioned in the description that a fix for another issue which is not yet merged will break this. So this issue will be relevant if the PR #833 is merged. It depends on the way we choose to fix issue #710: if i-- is performed for every iteration in the reverse mode, then we should account for the break case, where i++ was not performed in the original for loop, so the value should not be decremented before being read afterwards.

Parth wrote a [comment](https://github.com/vgvassilev/clad/pull/833#issuecomment-2018690457) on this #833, but I don't know if it still stands. This PR is blocked by #855 .

**Comment by vgvassilev** [comment]
#855 seems to be merged, is this issue resolved?

**Comment by kchristin22** [comment]
#833 tries to fix this issue, as well as the issue #710.
